Initial implementation of event retention on topics

hawkowl opened the pull request:
cc @oberstet 

This is a first approach going per https://github.com/crossbario/crossbar/issues/887#issuecomment-260326764 . I'm reusing the `extra` which seems to be None by default but nothing uses it -- could I change this to a `dict` by default? I can't find anything that touches it...

hawkowl commented:
Related AB bugfix I did whilst manually testing this: https://github.com/crossbario/autobahn-python/pull/755

hawkowl commented:
Using the hello example as a base:

```
            yield self.publish('com.example.oncounter', counter, options=PublishOptions(retain=True))
            self.log.info("published to 'oncounter' with counter {counter}",
                          counter=counter)
            counter += 1

            def oncount(msg, deets):
                print(deets)
                self.log.info("retained event for 'oncount' received: {msg}", msg=repr(msg))

            s = yield self.subscribe(oncount, 'com.example.oncounter', options=SubscribeOptions(get_retained=True, details_arg="deets"))

            yield sleep(1)
            yield s.unsubscribe()
```

I get output like this:

```
2016-11-15T23:25:15+1100 [Container   90404] published to 'oncounter' with counter 0
2016-11-15T23:25:15+1100 [Router      90403] Event(subscription=8648105260851741, publication=7264660483997918, args=[0], kwargs=None, publisher=None, publisher_authid=None, publisher_authrole=None, topic=None, retained=True, enc_algo=None, enc_key=None, enc_serializer=None, payload=-)
2016-11-15T23:25:15+1100 [Container   90404] EventDetails(publication=7264660483997918, publisher=None, publisher_authid=None, publisher_authrole=None, topic=<com.example.oncounter>, retained=True, enc_algo=None)
2016-11-15T23:25:15+1100 [Container   90404] retained event for 'oncount' received: 0
2016-11-15T23:25:16+1100 [Container   90404] published to 'oncounter' with counter 1
2016-11-15T23:25:16+1100 [Router      90403] Event(subscription=8648105260851741, publication=7714068808596375, args=[1], kwargs=None, publisher=None, publisher_authid=None, publisher_authrole=None, topic=None, retained=True, enc_algo=None, enc_key=None, enc_serializer=None, payload=-)
2016-11-15T23:25:16+1100 [Container   90404] EventDetails(publication=7714068808596375, publisher=None, publisher_authid=None, publisher_authrole=None, topic=<com.example.oncounter>, retained=True, enc_algo=None)
2016-11-15T23:25:16+1100 [Container   90404] retained event for 'oncount' received: 1
2016-11-15T23:25:17+1100 [Container   90404] published to 'oncounter' with counter 2
2016-11-15T23:25:17+1100 [Router      90403] Event(subscription=8648105260851741, publication=8666395629785886, args=[2], kwargs=None, publisher=None, publisher_authid=None, publisher_authrole=None, topic=None, retained=True, enc_algo=None, enc_key=None, enc_serializer=None, payload=-)
2016-11-15T23:25:17+1100 [Container   90404] EventDetails(publication=8666395629785886, publisher=None, publisher_authid=None, publisher_authrole=None, topic=<com.example.oncounter>, retained=True, enc_algo=None)
2016-11-15T23:25:17+1100 [Container   90404] retained event for 'oncount' received: 2
```

hawkowl commented:
@oberstet stuff I am not sure about: my duplication of the code https://github.com/crossbario/crossbar/pull/921/files#diff-8945283a6bce56d3d36c1032b21bcd78R461 and how it doesn't pass on the subscriber information (going always non-disclosed). I'm not sure if we should disclose it (because the publisher might have gone away?) but it might be useful to subscribers. What do you think? I also don't do anything about authids but the rest of the broker doesn't either, is that in a different spot?

oberstet commented on crossbar/router/broker.py:
```diff
@@ -86,7 +86,8 @@ def __init__(self, router, options=None):
-                                                      payload_encryption_cryptobox=True)
+                                                      payload_encryption_cryptobox=True,
+                                                      event_retention=True)
 
```
ordering

oberstet commented on crossbar/router/broker.py:
```diff
@@ -178,13 +179,19 @@ def processPublish(self, session, publish):
 
+        # check if the event is to be retained by inspecting the 'retain' flag
+        retain_event = False
+        if getattr(publish, "retain", False):
```
pls no dyn attrs .. must be `if publish.retain`

hawkowl replied:
Okay, cool -- just don't want to break master ('retain' is unreleased)

hawkowl replied:
lol the tests fail anyway because i did `subscribe.get_retained`
how do we work around this? release an AB?

oberstet commented on crossbar/router/broker.py:
```diff
@@ -224,6 +231,25 @@ def on_authorize_success(authorization):
+                        observation = self._subscription_map.get_observation(publish.topic, create_if_non_existing=True)
+
+                        if not observation.extra:
+                            observation.extra = {}
```
no, pls not a dict, but a class .. similar to how it is done here: https://github.com/crossbario/crossbar/blob/master/crossbar/router/dealer.py#L72 and the lines down inside the dealer ..

oberstet commented on crossbar/router/observation.py:
```diff
@@ -290,14 +290,16 @@ def add_observer(self, observer, uri, match=u"exact", extra=None, observer_extra
         return observation, was_already_observed, is_first_observer
 
-    def get_observation(self, uri, match=u"exact"):
+    def get_observation(self, uri, match=u"exact", create_if_non_existing=False):
```
no. "get_" methods should be idempotent and have no side effects

oberstet commented on crossbar/router/observation.py:
```diff
@@ -394,7 +404,7 @@ def get_observation_by_id(self, id):
         return self._observation_id_to_observation.get(id, None)
 
-    def drop_observer(self, observer, observation):
+    def drop_observer(self, observer, observation, delete=True):
```
whats that?

oberstet commented:
I think the main blocker is that we need to extend the `UriObservationMap` class ..

hawkowl commented:
@oberstet okay, I've changed them so that they aren't overloaded by keywords, and updated the places that use it. Could you take another look?

The builds will still fail as they don't have the right AB yet...

oberstet commented:
@hawkowl This looks good to me! It is touching quite some core parts, but this is necessary here. So, I'd merge, and see what happens;) Travis isn't happy yet though ..

hawkowl commented:
@oberstet it seems that 0.16.1 doesn't have the stuff we need...

```
$> python -m pip install -U autobahn
Collecting autobahn
  Downloading http://localhost:3141/root/pypi/+f/e36/b3584f94586f0/autobahn-0.16.1-py2.py3-none-any.whl (251kB)
    100% |████████████████████████████████| 256kB 40.4MB/s
Requirement already up-to-date: six>=1.10.0 in /Users/red/venvs/commands/lib/python2.7/site-packages (from autobahn)
Collecting txaio>=2.5.2 (from autobahn)
  Downloading http://localhost:3141/root/pypi/+f/a91/352fd91eb3887/txaio-2.5.2-py2.py3-none-any.whl
Installing collected packages: txaio, autobahn
  Found existing installation: txaio 2.5.1
    Uninstalling txaio-2.5.1:
      Successfully uninstalled txaio-2.5.1
  Found existing installation: autobahn 0.16.0
    Uninstalling autobahn-0.16.0:
      Successfully uninstalled autobahn-0.16.0
Successfully installed autobahn-0.16.1 txaio-2.5.2
You are using pip version 8.1.2, however version 9.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
$> python
Python 2.7.10 (default, Jul 30 2016, 18:31:42)
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.34)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from autobahn.wamp.message import Publish
>>> import inspect
>>> inspect.getargspec(Publish.__init__)
ArgSpec(args=['self', 'request', 'topic', 'args', 'kwargs', 'payload', 'acknowledge', 'exclude_me', 'exclude', 'exclude_authid', 'exclude_authrole', 'eligible', 'eligible_authid', 'eligible_authrole', 'enc_algo', 'enc_key', 'enc_serializer'], varargs=None, keywords=None, defaults=(None, None, None, None, None, None, None, None, None, None, None, None, None, None))
```

hawkowl commented:
That is, there's no retain argument. Maybe you didn't `git pull` before pushing the release, because the event retention stuff *did* come after you changed the version to 0.16.1... oh well, next version. I have more stuff to fit into Autobahn anyway :)

hawkowl commented:
@oberstet all-green is blocked on a new Autobahn release, it looks like...
